File: 剑指offer(Sword for offer)/复杂链表的复制.py
# -*- coding:utf-8 -*-

class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

# 方法1.我的方法。
# 方法1，我的方法。创建一个用于保存原链表-复制链表对应结点的字典
# 先遍历原链表，把next关系复制好。新建一个结点的同时，把原链表结点-复制链表结点键值对放到字典里
# 再遍历原链表，把random关系复制好。通过原链表-复制链表对应结点字典，找到random关系对应的结点
class Solution:
    # 返回 RandomListNode
    def Clone(self, h):
        # write code here
        if h == None:
            return None
        # 用于保存原链表-复制链表对应结点的字典
        node_address = {}
        copyH = RandomListNode(h.label)
        node_address[h] = copyH
        tmpH = h
        ansH = toolH = copyH
        # 先遍历原链表，把next关系复制好。新建一个结点的同时，把原链表结点-复制链表结点键值对放到字典里
        while tmpH.next != None:
            node = RandomListNode(tmpH.next.label)
            copyH.next = node
            node_address[tmpH.next] = node
            tmpH = tmpH.next
            copyH = copyH.next
        # 再遍历原链表，把random关系复制好。通过原链表-复制链表对应结点字典，找到random关系对应的结点
        while h != None:
            if h.random == None:
                toolH.random = None
            else:
                toolH.random = node_address[h.random]
            h = h.next
            toolH = toolH.next
            
        return ansH

# 方法2，思路：
    1. 把复制的结点链接在原始链表的每一对应结点后面
    2. 把复制的结点的random指针指向被复制结点的random指针的下一个结点
    3. 拆分成两个链表，奇数位置为原链表，偶数位置为复制链表，注意复制链表的最后一个结点的next指针不能跟原链表指向同一个空结点None，next指针要重新赋值None(判定程序会认定你没有完成复制）
    class Solution:
    # 返回 RandomListNode
    def Clone(self, pHead):
        if not pHead:
            return None
         
        dummy = pHead
         
        # first step, N' to N next
        while dummy:
            dummynext = dummy.next
            copynode = RandomListNode(dummy.label)
            copynode.next = dummynext
            dummy.next = copynode
            dummy = dummynext
         
        dummy = pHead
         
        # second step, random' to random'
        while dummy:
            dummyrandom = dummy.random
            copynode = dummy.next
            if dummyrandom:
                copynode.random = dummyrandom.next
            dummy = copynode.next
         
        # third step, split linked list
        dummy = pHead
        copyHead = pHead.next
        while dummy:
            copyNode = dummy.next
            dummynext = copyNode.next
            dummy.next = dummynext
            if dummynext:
                copyNode.next = dummynext.next
            else:
                copyNode.next = None
            dummy = dummynext
 
        return copyHead

File: 剑指offer(Sword for offer)/复杂链表的复制.cpp
// 输入一个复杂链表（每个节点中有节点值，以及两个指针，一个指向下一个节点，另一个特殊指针指向任意一个节点），返回结果为复制后复杂链表的head。
// （注意，输出结果中请不要返回参数中的节点引用，否则判题程序会直接返回空）

#include<iostream>
#include<vector>
#include<map>
using namespace std;

struct RandomListNode {
	int label;
	struct RandomListNode *next, *random;
	RandomListNode(int x) :
	label(x), next(NULL), random(NULL) {}
};

// 方法1，我的方法。创建一个用于保存原链表-复制链表对应结点的字典
// 先遍历原链表，把next关系复制好。新建一个结点的同时，把原链表结点-复制链表结点键值对放到字典里
// 再遍历原链表，把random关系复制好。通过原链表-复制链表对应结点字典，找到random关系对应的结点
class Solution {
public:
	RandomListNode* Clone(RandomListNode* h)
	{
		if(!h){return NULL;}
		// 用于保存原链表-复制链表对应结点的字典
		map<RandomListNode*, RandomListNode*> node_address; // node_address的key保存原链表的结点地址，value保存复制后链表的结点地址
		RandomListNode* copyH = new RandomListNode(h->label);
		node_address.insert(pair<RandomListNode*, RandomListNode*>(h, copyH));
		RandomListNode* tmpH = h;
		RandomListNode* ansH = copyH;
		RandomListNode* toolH = copyH;
		// 先遍历原链表，把next关系复制好。新建一个结点的同时，把原链表结点-复制链表结点键值对放到字典里
		while(tmpH->next){
			RandomListNode* tmpNext = new RandomListNode(tmpH->next->label);
			copyH->next = tmpNext;
			node_address.insert(pair<RandomListNode*, RandomListNode*>(tmpH->next, copyH->next));
			tmpH = tmpH->next;
			copyH = copyH->next;
		}
		// 再遍历原链表，把random关系复制好。通过原链表-复制链表对应结点字典，找到random关系对应的结点
		while(h){
			if(!h->random){
				toolH->random = NULL;
			}else{
				toolH->random = node_address[h->random];
			}
			h = h->next;
			toolH = toolH->next;
		}
		return ansH;
	}
};

// 方法2. 感觉没我的简便
/*
*解题思路：
*1、遍历链表，复制每个结点，如复制结点A得到A1，将结点A1插到结点A后面；
*2、重新遍历链表，复制老结点的随机指针给新结点，如A1.random = A.random.next;
*3、拆分链表，将链表拆分为原链表和复制后的链表
*/
class Solution {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if (!pHead) return NULL;
    nodeClone(pHead);
    connectRandom(pHead);
    return reconnect(pHead);
    }
     
     
//[1]复制结点，插入到原结点后方
void nodeClone(RandomListNode *head)
{
    RandomListNode *pNode = head;
    while (pNode != NULL)
    {
        RandomListNode *pClone = new RandomListNode(pNode->label);
        pClone->next = pNode->next;
        pNode->next = pClone;
        pNode = pClone->next;
    }
}
 
//[2]还原新结点的random指针
void connectRandom(RandomListNode *head)
{
    RandomListNode *pNode = head;
     
    while (pNode != NULL)
    {
        RandomListNode *pClone = pNode->next;
        if (pNode->random)
        {
            pClone->random = pNode->random->next;
        }
        pNode = pClone->next;
    }
}
 
//[3]拆分
RandomListNode *reconnect(RandomListNode *head)
{
    RandomListNode *pNode = head;
    RandomListNode *result = head->next;
    while (pNode != NULL)
    {
        RandomListNode *pClone = pNode->next;
        pNode->next = pClone->next;
        pNode = pNode->next;
        if (pNode != NULL)
            pClone->next = pNode->next;
         
    }
    return result;
}
 
};

